File: client/src/http_client.rs
use bitcoin::Txid;
use reqwest::{header::HeaderMap, Client as HttpClient, Response};
use std::{error::Error, str::FromStr};
use titan_types::{
    AddressData, Block, BlockTip, Pagination, PaginationResponse, RuneResponse, Status,
    Subscription, Transaction, TxOutEntry,
};

/// A client for your HTTP server.
pub struct Client {
    http_client: HttpClient,
    base_url: String,
}

impl Client {
    /// Create a new client with the given base URL (e.g. "http://localhost:3030").
    pub fn new(base_url: &str) -> Self {
        Self {
            http_client: HttpClient::new(),
            base_url: base_url.trim_end_matches('/').to_string(),
        }
    }

    // ------------------------------------------------------------------------
    // Status, blocks, addresses
    // ------------------------------------------------------------------------

    /// GET /status
    pub async fn get_status(&self) -> Result<Status, Box<dyn Error>> {
        let url = format!("{}/status", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        let status = resp.json::<Status>().await?;
        Ok(status)
    }

    /// GET /tip
    pub async fn get_tip(&self) -> Result<BlockTip, Box<dyn Error>> {
        let url = format!("{}/tip", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        let tip = resp.json::<BlockTip>().await?;
        Ok(tip)
    }

    /// GET /block/{query}
    pub async fn get_block(&self, query: &str) -> Result<Block, Box<dyn Error>> {
        let url = format!("{}/block/{}", self.base_url, query);
        let resp = self.http_client.get(&url).send().await?;
        let block = resp.json::<Block>().await?;
        Ok(block)
    }

    /// GET /address/{address}
    pub async fn get_address(&self, address: &str) -> Result<AddressData, Box<dyn Error>> {
        let url = format!("{}/address/{}", self.base_url, address);
        let resp = self.http_client.get(&url).send().await?;
        let address_data = resp.json::<AddressData>().await?;
        Ok(address_data)
    }

    // ------------------------------------------------------------------------
    // Transactions
    // ------------------------------------------------------------------------

    /// GET /tx/{txid}
    /// Returns a structured bitcoin `Transaction` (JSON).
    pub async fn get_transaction(
        &self,
        txid: &str,
    ) -> Result<bitcoin::Transaction, Box<dyn Error>> {
        let url = format!("{}/tx/{}", self.base_url, txid);
        let resp = self.http_client.get(&url).send().await?;
        let tx = resp.json::<bitcoin::Transaction>().await?;
        Ok(tx)
    }

    /// Same as `get_transaction` but always requests runes (`with_runes=true`).
    pub async fn get_transaction_with_runes(
        &self,
        txid: &str,
    ) -> Result<Transaction, Box<dyn Error>> {
        // Build query-string for the `with_runes` parameter.
        let url = format!("{}/tx/{}?with_runes=true", self.base_url, txid);
        let resp = self.http_client.get(&url).send().await?;
        let tx = resp.json::<Transaction>().await?;
        Ok(tx)
    }

    /// GET /tx/{txid}/raw
    /// Returns the raw transaction bytes (content-type: application/octet-stream).
    pub async fn get_transaction_raw(&self, txid: &str) -> Result<Vec<u8>, Box<dyn Error>> {
        let url = format!("{}/tx/{txid}/raw", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        // Read the raw bytes directly
        let bytes = resp.bytes().await?.to_vec();
        Ok(bytes)
    }

    /// GET /tx/{txid}/hex
    /// Returns a text/plain body containing hex-encoded raw transaction.
    pub async fn get_transaction_hex(&self, txid: &str) -> Result<String, Box<dyn Error>> {
        let url = format!("{}/tx/{txid}/hex", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        let hex = resp.text().await?; // text/plain
        Ok(hex)
    }

    /// POST /tx/broadcast
    /// Broadcast a raw transaction (in hex form) to the network.
    pub async fn broadcast_transaction(&self, tx_hex: String) -> Result<Txid, Box<dyn Error>> {
        let url = format!("{}/tx/broadcast", self.base_url);

        // 1) Build the request, consuming our owned `tx_hex`.
        let resp = self
            .http_client
            .post(&url)
            .body(tx_hex) // note: we pass the owned string here
            .send()
            .await?;

        // 2) Extract status and body text. We can't read `resp.text()` twice,
        //    so we store them now before checking the status.
        let status = resp.status();
        let body_text = resp.text().await?;

        // 3) If not successful, return an error with the body as context.
        if !status.is_success() {
            return Err(format!("Broadcast failed: HTTP {} - {}", status, body_text).into());
        }

        // 4) At this point, we know it's success. The response body is the txid as text.
        let txid = Txid::from_str(&body_text)?;
        Ok(txid)
    }

    /// GET /output/{outpoint}
    pub async fn get_output(&self, outpoint: &str) -> Result<TxOutEntry, Box<dyn Error>> {
        let url = format!("{}/output/{}", self.base_url, outpoint);
        let resp = self.http_client.get(&url).send().await?;
        let output = resp.json::<TxOutEntry>().await?;
        Ok(output)
    }

    // ------------------------------------------------------------------------
    // Inscriptions
    // ------------------------------------------------------------------------

    /// GET /inscription/{inscription_id}
    /// Returns `(Headers, Bytes)`. The server sets content-type and possibly compression,
    /// and the body is the raw inscription content.
    pub async fn get_inscription(
        &self,
        inscription_id: &str,
    ) -> Result<(HeaderMap, Vec<u8>), Box<dyn Error>> {
        let url = format!("{}/inscription/{}", self.base_url, inscription_id);
        let resp: Response = self.http_client.get(&url).send().await?;

        let status = resp.status();

        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(format!("Inscription request failed: HTTP {} - {}", status, body).into());
        }
        let headers = resp.headers().clone();
        let body = resp.bytes().await?.to_vec();
        Ok((headers, body))
    }

    // ------------------------------------------------------------------------
    // Runes
    // ------------------------------------------------------------------------

    /// GET /runes?skip=&limit=
    pub async fn get_runes(
        &self,
        pagination: Option<Pagination>,
    ) -> Result<PaginationResponse<RuneResponse>, Box<dyn Error>> {
        let url = format!("{}/runes", self.base_url);
        let mut req = self.http_client.get(&url);
        if let Some(pagination) = pagination {
            req = req.query(&[("skip", pagination.skip), ("limit", pagination.limit)]);
        }
        let resp = req.send().await?;
        let runes = resp.json::<PaginationResponse<RuneResponse>>().await?;
        Ok(runes)
    }

    /// GET /rune/{rune}
    pub async fn get_rune(&self, rune: &str) -> Result<RuneResponse, Box<dyn Error>> {
        let url = format!("{}/rune/{}", self.base_url, rune);
        let resp = self.http_client.get(&url).send().await?;
        let rune_response = resp.json::<RuneResponse>().await?;
        Ok(rune_response)
    }

    /// GET /rune/{rune}/transactions?skip=&limit=
    pub async fn get_rune_transactions(
        &self,
        rune: &str,
        pagination: Option<Pagination>,
    ) -> Result<PaginationResponse<Txid>, Box<dyn Error>> {
        let url = format!("{}/rune/{}/transactions", self.base_url, rune);
        let mut req = self.http_client.get(&url);
        if let Some(pagination) = pagination {
            req = req.query(&[("skip", pagination.skip), ("limit", pagination.limit)]);
        }
        let resp = req.send().await?;
        let transactions = resp.json::<PaginationResponse<Txid>>().await?;
        Ok(transactions)
    }

    /// GET /mempool/txids
    pub async fn get_mempool_txids(&self) -> Result<Vec<Txid>, Box<dyn Error>> {
        let url = format!("{}/mempool/txids", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        let txids = resp.json::<Vec<Txid>>().await?;
        Ok(txids)
    }

    // ------------------------------------------------------------------------
    // Subscription endpoints
    // ------------------------------------------------------------------------

    /// GET /subscription/{id}
    pub async fn get_subscription(&self, id: &str) -> Result<Subscription, Box<dyn Error>> {
        let url = format!("{}/subscription/{}", self.base_url, id);
        let resp = self.http_client.get(&url).send().await?;
        let subscription = resp.json::<Subscription>().await?;
        Ok(subscription)
    }

    /// GET /subscriptions
    pub async fn list_subscriptions(&self) -> Result<Vec<Subscription>, Box<dyn Error>> {
        let url = format!("{}/subscriptions", self.base_url);
        let resp = self.http_client.get(&url).send().await?;
        let subscriptions = resp.json::<Vec<Subscription>>().await?;
        Ok(subscriptions)
    }

    /// POST /subscription
    pub async fn add_subscription(
        &self,
        subscription: &Subscription,
    ) -> Result<Subscription, Box<dyn Error>> {
        let url = format!("{}/subscription", self.base_url);
        let resp = self
            .http_client
            .post(&url)
            .json(subscription)
            .send()
            .await?;
        let added = resp.json::<Subscription>().await?;
        Ok(added)
    }

    /// DELETE /subscription/{id}
    pub async fn delete_subscription(&self, id: &str) -> Result<(), Box<dyn Error>> {
        let url = format!("{}/subscription/{}", self.base_url, id);
        let resp = self.http_client.delete(&url).send().await?;
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        if !status.is_success() {
            return Err(format!("Delete subscription failed: HTTP {} - {}", status, body).into());
        }
        Ok(())
    }
}

File: ts-sdk/src/http-client.ts
import axios, { AxiosInstance, AxiosError, AxiosRequestConfig } from 'axios';
import {
  AddressData,
  BlockTip,
  Pagination,
  PaginationResponse,
  RuneResponse,
  Subscription,
  Status,
  Transaction,
  TxOutEntry,
  TransactionStatus,
  Block,
  MempoolEntry,
} from './types';

/**
 * A HTTP client class for the Titan API. Each method corresponds
 * to an endpoint (for example: GET /status, GET /tx/:txid, POST /tx/broadcast, etc.)
 */
export class TitanHttpClient {
  private http: AxiosInstance;
  private baseUrl: string;

  constructor(baseUrl: string) {
    // Remove any trailing slashes from the URL.
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.http = axios.create({
      baseURL: this.baseUrl,
    });
  }

  async getStatus(): Promise<Status> {
    return await this.getOrFail<Status>('/status');
  }

  async getTip(): Promise<BlockTip> {
    return await this.getOrFail<BlockTip>('/tip');
  }

  /**
   * Fetches a block by its query (could be a block height or hash).
   */
  async getBlock(query: string): Promise<Block | undefined> {
    return await this.get<Block>(`/block/${query}`);
  }

  async getBlockHashByHeight(height: number): Promise<string | undefined> {
    return await this.get<string>(`/block/${height}/hash`);
  }

  async getBlockTxids(query: string): Promise<string[] | undefined> {
    return await this.get<string[]>(`/block/${query}/txids`);
  }

  async getAddress(address: string): Promise<AddressData> {
    return await this.getOrFail<AddressData>(`/address/${address}`);
  }

  async getTransaction(txid: string): Promise<Transaction | undefined> {
    return await this.get<Transaction>(`/tx/${txid}`);
  }

  async getTransactionRaw(txid: string): Promise<Uint8Array | undefined> {
    // Request raw binary data using the arraybuffer responseType.
    const response = await this.get<ArrayBuffer>(`/tx/${txid}/raw`, {
      responseType: 'arraybuffer',
    });

    if (!response) {
      return undefined;
    }

    return new Uint8Array(response);
  }

  async getTransactionHex(txid: string): Promise<string | undefined> {
    return await this.get<string>(`/tx/${txid}/hex`);
  }

  async getTransactionStatus(
    txid: string,
  ): Promise<TransactionStatus | undefined> {
    return await this.get<TransactionStatus>(`/tx/${txid}/status`);
  }

  async sendTransaction(txHex: string): Promise<string> {
    try {
      const response = await this.http.post<string>('/tx/broadcast', txHex, {
        headers: {
          'Content-Type': 'text/plain',
        },
      });

      return response.data;
    } catch (err) {
      if (err instanceof AxiosError) {
        const axiosError = err as AxiosError;
        if (axiosError.response?.data) {
          const error = axiosError.response.data as any;
          throw error;
        }
      }

      throw err;
    }
  }

  async getOutput(txid: string, vout: number): Promise<TxOutEntry | undefined> {
    return await this.get<TxOutEntry>(`/output/${txid}:${vout}`);
  }

  async getInscription(
    inscriptionId: string,
  ): Promise<{ headers: any; data: Uint8Array }> {
    const response = await this.http.get<ArrayBuffer>(
      `/inscription/${inscriptionId}`,
      {
        responseType: 'arraybuffer',
      },
    );

    return {
      headers: response.headers,
      data: new Uint8Array(response.data),
    };
  }

  async getRunes(
    pagination?: Pagination,
  ): Promise<PaginationResponse<RuneResponse>> {
    const params = pagination || {};
    return await this.getOrFail<PaginationResponse<RuneResponse>>('/runes', {
      params,
    });
  }

  async getRune(rune: string): Promise<RuneResponse | undefined> {
    return await this.get<RuneResponse>(`/rune/${rune}`);
  }

  async getRuneTransactions(
    rune: string,
    pagination?: Pagination,
  ): Promise<PaginationResponse<string> | undefined> {
    const params = pagination || {};
    return await this.getOrFail<PaginationResponse<string>>(
      `/rune/${rune}/transactions`,
      { params },
    );
  }

  async getMempoolTxids(): Promise<string[]> {
    return await this.getOrFail<string[]>(`/mempool/txids`);
  }

  async getMempoolEntry(txid: string): Promise<MempoolEntry | undefined> {
    return await this.get<MempoolEntry>(`/mempool/entry/${txid}`);
  }

  async getMempoolEntries(
    txids: string[],
  ): Promise<Map<string, MempoolEntry | undefined>> {
    const response = await this.http.post<
      Record<string, MempoolEntry | undefined>
    >('/mempool/entries', txids);

    return new Map(Object.entries(response.data));
  }

  async getAllMempoolEntries(): Promise<Map<string, MempoolEntry>> {
    const response = await this.http.get<Record<string, MempoolEntry>>(
      '/mempool/entries/all',
    );

    return new Map(Object.entries(response.data));
  }

  async getSubscription(id: string): Promise<Subscription | undefined> {
    return await this.get<Subscription>(`/subscription/${id}`);
  }

  async listSubscriptions(): Promise<Subscription[]> {
    return await this.getOrFail<Subscription[]>('/subscriptions');
  }

  async addSubscription(subscription: Subscription): Promise<Subscription> {
    const response = await this.http.post<Subscription>(
      '/subscription',
      subscription,
    );
    return response.data;
  }

  async deleteSubscription(id: string): Promise<void> {
    const response = await this.http.delete(`/subscription/${id}`);
    if (response.status < 200 || response.status >= 300) {
      throw new Error(`Delete subscription failed: HTTP ${response.status}`);
    }
  }

  private async getOrFail<T>(
    path: string,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    const response = await this.http.get<T>(path, config);
    return response.data;
  }

  private async get<T>(
    path: string,
    config?: AxiosRequestConfig,
  ): Promise<T | undefined> {
    try {
      const response = await this.http.get<T>(path, config);
      return response.data;
    } catch (err) {
      if (err instanceof AxiosError) {
        const axiosError = err as AxiosError;
        if (axiosError.response?.status === 404) {
          return undefined;
        }
      }

      throw err;
    }
  }
}
